core/services/pipeline/orm.go

```diff
@@ -542,3 +601,76 @@ func loadAssociations(q pg.Queryer, runs []*Run) error {
 func (o *orm) GetQ() pg.Q {
 	return o.q
 }
+
+func (o *orm) loadCount(pipelineSpecID int32) *atomic.Uint64 {
+	actual, _ := o.pm.LoadOrStore(pipelineSpecID, new(atomic.Uint64))
+	return actual.(*atomic.Uint64)
+}
+
+// Prune attempts to keep the pipeline_runs table capped close to the
+// maxSuccessfulRuns length for each pipeline_spec_id.
+//
+// It does this synchronously for small values and async/sampled for large
+// values.
+//
+// Note this does not guarantee the pipeline_runs table is kept to exactly the
+// max length, rather that it doesn't excessively larger than it.
+func (o *orm) Prune(tx pg.Queryer, pipelineSpecID int32) {
+	if pipelineSpecID == 0 {
+		logger.Sugared(o.lggr).AssumptionViolation("Cannot prune pipeline_runs: expected a non-zero pipeline spec ID")
```

Comment:
AssumptionViolation just logs with a prefix at error level. Should this also return? Or be elevated to Panic?

Reply:
Elevating to panic :+1: